tests/mbt/driver/mbt_test.go

```diff
@@ -0,0 +1,577 @@
+		consumerCurValSet := make(map[string]int64, len(consumerVals))
+		for _, val := range consumerVals {
+			pubkey, err := val.ConsPubKey()
+			require.NoError(t, err, "Error getting pubkey")
+
+			consAddr := providertypes.NewConsumerConsAddress(sdktypes.ConsAddress(pubkey.Address().Bytes()))
+
+			providerConsAddr, found := driver.providerKeeper().GetValidatorByConsumerAddr(driver.providerCtx(), consumer, consAddr)
+			if !found {
+				providerConsAddr = providertypes.NewProviderConsAddress(consAddr.Address)
+			}
+
+			providerVal, found := driver.providerStakingKeeper().GetValidatorByConsAddr(driver.providerCtx(), providerConsAddr.Address)
+			require.True(t, found, "Error getting provider validator")
+
+			consumerCurValSet[providerVal.GetMoniker()] = val.Power
+		}
+		for val, power := range modelValSet {
+			_ = val
+			intPow := power.Value.(int64)
+			_ = intPow
+		}
+		require.NoError(t, CompareValSet(modelValSet, consumerCurValSet), "Validator sets do not match for consumer %v", consumer)
+	}
+}
+
+// ComparePacketQueues compares the packet queues in the model to the packet queues in the system.
+// It compares both incoming (provider->consumer) and outgoing (consumer->provider) packets.
+// It only takes the number of packets into account, not the contents.
+func ComparePacketQueues(
+	t *testing.T,
+	driver *Driver,
+	currentModelState map[string]itf.Expr,
+	consumer string,
+	timeOffset time.Time,
+) {
+	t.Helper()
+	ComparePacketQueue(t, driver, currentModelState, Provider, consumer, timeOffset)
+	ComparePacketQueue(t, driver, currentModelState, consumer, Provider, timeOffset)
+}
+
+func ComparePacketQueue(
+	t *testing.T,
+	driver *Driver,
+	currentModelState map[string]itf.Expr,
+	sender string,
+	receiver string,
+	timeOffset time.Time,
+) {
+	t.Helper()
+	modelSenderQueue := PacketQueue(currentModelState, sender, receiver)
+	actualSenderQueue := driver.packetQueue(ChainId(sender), ChainId(receiver))
+
+	require.Equal(t,
+		len(modelSenderQueue),
+		len(actualSenderQueue),
+		"Packet queue sizes do not match for sender %v, receiver %v",
+		sender,
+		receiver)
+
+	for i := range modelSenderQueue {
+		actualPacket := actualSenderQueue[i]
+		modelPacket := modelSenderQueue[i].Value.(itf.MapExprType)
+
+		actualTimeout := time.Unix(0, int64(actualPacket.Packet.TimeoutTimestamp)).Unix() - timeOffset.Unix()
+		modelTimeout := GetTimeoutForPacket(modelPacket)
+
+		require.Equal(t,
+			modelTimeout,
+			actualTimeout,
+			"Timeouts do not match for packet %v, sender %v, receiver %v",
+			actualPacket,
+			sender,
+			receiver)
+	}
+}
+
+// CompareTimes compares the block times in the model to the block times in the system.
+// It takes into account the timeOffsets, which should be the starting times
+// of the chains in the system after the system has been initialized.
+// We only compare down to seconds, because the model and system will differ
+// on the order of nanoseconds.
```

Comment:
I guess rare but still possible that one time would be `second X nanonseconds 999999999 ` and the other `second X + 1 nanoseconds 0`. In such a case the comparison would fail, right? Would it make sense to compare their absolute value instead and verify that it's `<= 100ms` or something in this vein?

Reply:
I think I like it better as-is - I want to avoid the impression that it's just that they differ on the order of hundreds of milliseconds.
I need to increment timestamps between blocks by nanoseconds more than the model, but only in some specific cases. Rounding down is always more "correct" than checking some arbitrary difference, because it will *always* be that the model has time `second X` and the actual system has time `second X nanoseconds Y` until it spills over into seconds.

Hope that makes some amount of sense? I wrote a bit of the explanation into the comment

Reply:
I see. Sounds good. Thanks for the explanation. 😃